### DISCUSSION TITLE
Is it possible to mux VVC bitstream into a container with audio already?

### DISCUSSION THREAD

**Discussion by Fujishima-Rurino** [fraunhoferhhi/vvenc#403] [General] [answered]
I'm experimenting with VVC to see how big the compression gains without perceptible quality loss are on some of the h264/VP9 1080p encodes I have and I've tried muxing the results into mp4/mkv, but so far I have trouble playing that back properly on my system. 

Here's the bitstream and audio:
[UtsukushikuIkiro-VID](https://github.com/user-attachments/files/16227596/UtsukushikuIkiro.zip)
[UtsukushikuIkiro-AUD](https://github.com/user-attachments/files/16227601/UtsukushikuIkiro-AUD.zip)

Playing back the muxed mp4 on MPC-HC, it stutters terribly and is virtually unwatchable due to this, while mkv cannot be reliably muxed at all in ffmpeg (timestamp creation anyone?).
The one solution that I found is rather cumbersome and involves muxing on the fly into a NUT container:

```
ffmpeg -strict -2 -r 24000/1001 -i UtsukushikuIkiro.h266 -i UtsukushikuIkiro.opus -c:v rawvideo -pix_fmt yuv420p10le -s 1920x1080 -c:a copy -f nut - | ffplay -f nut -i -
``` 

That doesn't stutter at all, doesn't eat up unreasonable amounts of RAM and has decent enough visual quality. Ideally, though, you don't have to do all that and can just play a muxed container. Also, YUV conversion does introduce a slight color tint here.

Is it really overhead of demuxing that causes so much stutter? I'm amazed with visual quality I get from MPC-HC on playing back standalone VVC bitsream so it would be lovely if I could somehow get it to play back the muxed container stutter-free.

**Comment by adamjw24** [comment]
How do you MUX to mp4? This should be the solution to use. We have been doing it since forever and it works great for us. If using GPAC it is recommended to use -forcesync (or whatever the actualy syntax is) to enable seeking in with open GOP RAPs.

AFAIK mkv with VVC is not specified or implemented and generally does not work.

**Comment by lehmann-c** [comment, marked as answer]
By using MP4Box from [gpac](https://gpac.io) it´s pretty easy to mux audio and video:
`MP4Box -add UtsukushikuIkiro.h266 -add UtsukushikuIkiro.opus -new UtsukushikuIkiro.mp4`  
or  
`MP4Box -add UtsukushikuIkiro.h266 -add UtsukushikuIkiro.opus -new UtsukushikuIkiro.mkv`

**Comment by Fujishima-Rurino** [comment]
Oh man, MP4Box does work like charm, thanks a lot! mp4 and even mkv play smoothly in MPC-HC. 

I was muxing with latest ffmpeg and apparently it could never put sensible PTS in the generated file. There's probably a way to get it right even with their current level of support eventually with bitstream filters, but no thanks, I'll take GPAC for this one.
